【wc コマンドを作る】プラクティス課題

shugoStr opened the pull request:
提出いたします。
ご確認よろしくお願いいたします！

shugoStr committed:
first commit

before refactoring

shugoStr committed:
Implement Main class

shugoStr committed:
Implement class StanderdInput and FileArguments

shugoStr committed:
refactoring: Implement decisions in opt.

shugoStr committed:
Implement class CalcFileInfo and display methods

shugoStr committed:
add method:calc_bytesize

JunichiIto reviewed (commented):
動画でレビューしたので別途リンクを送ります〜。

shugoStr committed:
Importing reviews by JunichiIto

Consolidate to Main class.Overall revision

shugoStr committed:
Implement render_counted_info

The method function is return the result as an array. Thne fix the RUBOCOP point.

shugoStr committed:
Revisit and correct back

About RUBOCOP errors

JunichiIto commented:
今回も動画でレビューしたので別途リンクを送ります〜。

shugoStr committed:
Fixed a fatal bug. About total calculation error.

Then importing Ito's review.

JunichiIto commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,81 @@
+      line_count, word_count, bytesize_count = count_file_info(file_text)
+      display_detailed_info(line_count, word_count, bytesize_count, opt)
+      total_line += line_count
+      unless opt['l']
```
条件分岐無しでとりあえずwordやbytesizeも加算して、`display_total_count`で出すか出さないかを決める、というシンプルな実装にするのもありかも。
数値の加算ぐらいであればそこまでパフォーマンスにも影響ないでしょうし。

JunichiIto commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,81 @@
+    [line_count, word_count, bytesize_count]
+  end
+
+  def display_detailed_info(line_count, word_count, bytesize_count, opt)
```
19行目や35行目のputsもこのメソッドで担当できるように、引数を増やすといいかもしれません。

JunichiIto commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,81 @@
+    file_text.bytesize
+  end
+
+  def display_total_count(total_line, total_word, total_bytesize, opt)
```
`display_detailed_info`と実装がほぼ同じなので、メソッドの統合を検討してみてください。

JunichiIto commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,81 @@
+    display_total_count(total_line, total_word, total_bytesize, opt) if file_paths.size > 1
+  end
+
+  def count_file_info(file_text)
```
日報で「ツッコまれそう」と書いていましたが、意図はずいぶん伝わりやすくなりましたし、さらに良い命名もぱっと思いつかないので、これで良いと思います〜。

shugoStr committed:
Integrating methods. Add arguments and change logic.

Importing Ito's review.

shugoStr committed:
Add keyword argument.

About #display_detailed_info

JunichiIto commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,69 @@
+    display_detailed_info(line_count: total_line, word_count: total_word, bytesize_count: total_bytesize, opt: opt, file_path: 'total') if file_paths.size > 1
+  end
+
+  def display_detailed_info(line_count:, word_count:, bytesize_count:, opt:, file_path: nil)
```
https://github.com/shugoStr/ruby-practices/pull/9/commits/f929f0dfdbc8ec55781a884a519d535e1d35ad7f でキーワード引数を導入していますが、メソッド呼び出し時にちょっと横に長くなりすぎるので個人的にはキーワード引数を使わない方が良かったんじゃないかな〜と思いました。

ただ、その場合も、

```ruby
def display_detailed_info(line_count, word_count, bytesize_count, file_path, opt)
```

こんなふうに`opt`は一番最後に持ってくる方が、収まりがいい気がします（`opt`はオプションであって主役ではないので）。

JunichiIto commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,69 @@
+
+  private
+
+  def display_normal_input(opt)
```
細かいですが、標準入力を対象にするなら技術用語的に`stdin`とか`standard_input`とした方が誤解が少なくなりそうです。

ただ、`display_stdin`だと「標準入力を表示する」みたいな意味に取れてしまうので、`display_for_stdin`とか`display_for_file_paths`みたいな名前の方がいいかなと思います。

JunichiIto reviewed (approved):
2点ほどコメントしましたが、いずれも致命的な問題ではないので、これで合格とします〜 🎉

shugoStr committed:
fix method arguments and name

display_for_file_paths and display_for_stdin

shugoStr committed:
fix

Remove unwanted space.

shugoStr commented:
LGTM!
マージします！
